### PR TITLE
refactoring: idempotent reconciler (PROJQUAY-2046)

### PR DESCRIPTION
**Pull request by ricardomaraschini**
Allowing Reconciler to run multiple times without generating side effects.
    
- Stop creating new secrets every time
- Stop running database upgrade jobs every time
- Stop creating a job to run database init job (enable extension)
- Stop panicking in some places (returning errors instead)
- Cleaning logging string (stop logging a ton of uneeded info)
- Small improvements everywhere
- Stop setting number of replicas on deployments
    - They were overridding whatever was being set by hpa
- Removed unused functions and constants

### CODE REVIEW

File: kustomize/base/quay.deployment.yaml
Comment (ibazulic) [inline kustomize/base/quay.deployment.yaml:10]: This will once again change how the operator deploys Quay when HPA is disabled. We will once again be deploying 1 pod, while HPA enabled deployments deploy 2. We specifically changed this because customers asked for at least two pods every time when Quay is deployed via the operator.

File: controllers/quay/features.go
Comment (flavianmissi) [inline controllers/quay/features.go]: This feels like an error, but the message is logged as info, and no error is returned. What am I missing?

File: controllers/quay/features.go
Comment (flavianmissi) [inline controllers/quay/features.go:229]: Should we also log this message on line 229? `"unable to retrieve Quay datastore Secret"` is quite cryptic on its own...

File: controllers/quay/features.go
Comment (flavianmissi) [inline controllers/quay/features.go:235]: ditto

File: controllers/quay/features.go
Comment (flavianmissi) [inline controllers/quay/features.go]: s/checkRoutesAvailable/checkBuildManagerAvailable
s/etry/entry

File: controllers/quay/features.go
Comment (flavianmissi) [inline controllers/quay/features.go]: Don't we already log this as an error on the caller's side?

**Review by flavianmissi — COMMENTED** [review]
left some comments/questions, but overall it's looking good!

File: controllers/quay/features.go
Comment (ricardomaraschini) [reply controllers/quay/features.go]: I kept this as is to maintain the behavior already in place. That being said I think it makes sense to return an error here (this will make the reconcile loop to fail and a proper status to be recorded). I will fix it.

File: controllers/quay/features.go
Comment (ricardomaraschini) [reply controllers/quay/features.go:229]: I don't follow, what do you mean?

File: controllers/quay/features.go
Comment (ricardomaraschini) [reply controllers/quay/features.go:235]: I don't follow, what do you mean?

File: controllers/quay/features.go
Comment (ricardomaraschini) [reply controllers/quay/features.go]: done.

File: controllers/quay/features.go
Comment (ricardomaraschini) [reply controllers/quay/features.go]: Removing the log.

File: controllers/quay/features.go
Comment (flavianmissi) [reply controllers/quay/features.go:229]: When this happens, the logs will show an error saying "unable to retrieve Quay datastore `Secret`". That alone doesn't say much.
My suggestion would be to add something like the returned error's message to the log, i.e:
```
r.Log.Error(err, "unable to retrieve Quay datastore `Secret` - bucket claim not ready")
```
Or something in the likes.

Comment (ricardomaraschini) [reply controllers/quay/features.go:229]: Fixed, PTAL.

File: controllers/quay/features.go
Comment (ricardomaraschini) [reply controllers/quay/features.go:235]: Fixed, PTAL.

**Review by flavianmissi — DISMISSED** [review]
LGTM

**Comment by ricardomaraschini** [comment]
> LGTM

Thanks. I am going to wait for @jonathankingfc review as well as this is too big.

File: controllers/quay/features.go
Comment (jonathankingfc) [inline controllers/quay/features.go:91]: Could these values potentially be nil? I think we can add the ok syntax to the map check in lines 88 and 89.

File: kustomize/base/quay.deployment.yaml
Comment (jonathankingfc) [inline kustomize/base/quay.deployment.yaml:10]: Is there a specific reason for lowering the amount of replicas?

File: kustomize/components/clair/clair.deployment.yaml
Comment (jonathankingfc) [inline kustomize/components/clair/clair.deployment.yaml:10]: Same as above comment

**Review by jonathankingfc — CHANGES_REQUESTED** [review]

File: controllers/quay/features.go
Comment (ricardomaraschini) [reply controllers/quay/features.go:91]: I don't think this is needed. If indexes `ssl.cert` and `ssl.key` don't exist then `providedTLSCert` and `providedTLSKey` will be `nil` and `len()` of a `nil` slice is zero. Did I follow you reasoning or I missed something ?

**Review by ibazulic — CHANGES_REQUESTED** [review]
This will once again change how the operator deploys Quay when HPA is disabled. We will once again be deploying 1 pod, while HPA enabled deployments deploy 2. We specifically changed this because customers asked for at least two pods every time when Quay is deployed via the operator.

File: kustomize/base/quay.deployment.yaml
Comment (ricardomaraschini) [reply kustomize/base/quay.deployment.yaml:10]: If not removed then we will have a concurrency problem with the HPA as we can't assume the number of replicas set by the HPA (assume for instance that the HPA set this field to `3`, we can't reset it back to `2`). i.e. if we don't set it keeps the value set by the HPA.

Comment (ricardomaraschini) [reply kustomize/base/quay.deployment.yaml:10]: I don't think that to have a fixed value here applies to all scenarios. Isn't it better to have this working through an override?

Comment (ibazulic) [reply kustomize/base/quay.deployment.yaml:10]: override through CRD would be neat :-)

Comment (ricardomaraschini) [reply kustomize/base/quay.deployment.yaml:10]: As agreed I added a new commit to this PR that resets the number of replicas to 2 if HPA is unmanaged. As soon as this one gets merged we can work in having the value of replicas set through an override in CRD. Can you resolve this one if you are OK with it?

https://github.com/quay/quay-operator/pull/623/commits/f9b0b28fc40975940c93d47d375c53ec2a507b90

Comment (ricardomaraschini) [reply kustomize/base/quay.deployment.yaml:10]: After discussing this with @ibazulic I have restored the old behavior with a small difference: only set the number of replicas to 2 if HPA is not managed. Can you resolve this if you are OK with it?

**Review by flavianmissi — APPROVED** [review]

**Review by jonathankingfc — APPROVED** [review]
LGTM

**Comment by flavianmissi** [comment]
Do we want to backport this to `redhat-3.6` @ricardomaraschini ?

**Comment by ricardomaraschini** [comment]
> Do we want to backport this to `redhat-3.6` @ricardomaraschini ?

@flavianmissi no. Better to keep it in 3.7 (too many changes to include in a minor). I initially thought it would be good to backport but due to the size I reconsidered.

**Pull request merged**
